File: packages/games/arena/logic/src/ray.ts
import { OCCLUDERS } from './arena.ts';
import type { Seat } from './arena.ts';
import type { Bounds } from './bounds.ts';
import { MAX_SHOT_DISTANCE } from './constants.ts';
import type { BodyPart } from './pose.ts';
import type { Vec3 } from './vector.ts';

/**
 * Where a shot goes, and what it finds.
 *
 * This is in the shared rules rather than on the server alone, and that was a
 * deliberate change of mind. Hit registration is the most disputed mechanic in
 * any shooter; leaving it as the one rule the conformance vectors did not pin
 * would have been leaving out the part most worth pinning. It costs one
 * carefully written slab test.
 */

/**
 * How far along the ray it enters the box, or nothing if it never does.
 *
 * The classic form of this multiplies by `1 / direction` and lets the
 * infinities sort themselves out. That form produces `0 × ∞ = NaN` for a ray
 * running parallel to a face and starting on its plane, and Go and TypeScript
 * do not agree about how a NaN travels through the comparisons that follow. So
 * a degenerate axis is branched on and never divided through.
 *
 * A ray starting inside the box enters it at zero.
 */
export function rayVsBox(
  origin: Vec3,
  direction: Vec3,
  box: Bounds,
  maxDistance: number,
): number | null {
  let enter = 0;
  let exit = maxDistance;

  if (direction.x === 0) {
    if (origin.x < box.minX || origin.x > box.maxX) {
      return null;
    }
  } else {
    const a = (box.minX - origin.x) / direction.x;
    const b = (box.maxX - origin.x) / direction.x;
    const near = a < b ? a : b;
    const far = a < b ? b : a;
    if (near > enter) {
      enter = near;
    }
    if (far < exit) {
      exit = far;
    }
    if (enter > exit) {
      return null;
    }
  }

  if (direction.y === 0) {
    if (origin.y < box.minY || origin.y > box.maxY) {
      return null;
    }
  } else {
    const a = (box.minY - origin.y) / direction.y;
    const b = (box.maxY - origin.y) / direction.y;
    const near = a < b ? a : b;
    const far = a < b ? b : a;
    if (near > enter) {
      enter = near;
    }
    if (far < exit) {
      exit = far;
    }
    if (enter > exit) {
      return null;
    }
  }

  if (direction.z === 0) {
    if (origin.z < box.minZ || origin.z > box.maxZ) {
      return null;
    }
  } else {
    const a = (box.minZ - origin.z) / direction.z;
    const b = (box.maxZ - origin.z) / direction.z;
    const near = a < b ? a : b;
    const far = a < b ? b : a;
    if (near > enter) {
      enter = near;
    }
    if (far < exit) {
      exit = far;
    }
    if (enter > exit) {
      return null;
    }
  }

  return enter;
}

/**
 * How far along the ray it enters a box with an orientation of its own.
 *
 * The same slab test, done in the box's frame instead of the world's. A part of
 * a body carries its own right, up and forward, so the ray's origin and
 * direction are projected onto those three axes with dot products and the
 * result is an axis-aligned problem again — one that reads from `-half` to
 * `+half` on each axis.
 *
 * Dot products and the existing test: no matrix, no inverse, and nothing that
 * is not a multiply and an add.
 */
export function rayVsOrientedBox(
  origin: Vec3,
  direction: Vec3,
  box: OrientedBox,
  maxDistance: number,
): number | null {
  const dx = origin.x - box.centre.x;
  const dy = origin.y - box.centre.y;
  const dz = origin.z - box.centre.z;
  const along = (axis: Vec3): number => dx * axis.x + dy * axis.y + dz * axis.z;
  const cast = (axis: Vec3): number =>
    direction.x * axis.x + direction.y * axis.y + direction.z * axis.z;

  return rayVsBox(
    { x: along(box.right), y: along(box.up), z: along(box.forward) },
    { x: cast(box.right), y: cast(box.up), z: cast(box.forward) },
    {
      minX: -box.half.x,
      minY: -box.half.y,
      minZ: -box.half.z,
      maxX: box.half.x,
      maxY: box.half.y,
      maxZ: box.half.z,
    },
    maxDistance,
  );
}

/** A box that is not aligned to the world: a limb, a torso, a rifle. */
export interface OrientedBox {
  readonly centre: Vec3;
  readonly half: Vec3;
  readonly right: Vec3;
  readonly up: Vec3;
  readonly forward: Vec3;
}

/** One part of one body, where it stood at the judged moment. */
export interface ShotTarget {
  readonly seat: Seat;
  readonly part: BodyPart;
  readonly box: OrientedBox;
}

export interface Trace {
  /** Who was hit, or nobody. */
  readonly hitSeat: Seat | null;
  /** Which part of them, which is what decides the damage. */
  readonly hitPart: BodyPart | null;
  /** Where the shot stopped: a body, a wall, or the end of its range. */
  readonly endpoint: Vec3;
  readonly distance: number;
}

/**
 * Fires one shot and reports what it found.
 *
 * Walls are tested before bodies and win ties, so a target standing exactly
 * against the far side of a crate is behind it rather than in front of it.
 * Targets are given as boxes rather than as bodies because the server hands
 * over where they *were* when the shooter saw them, not where they are now.
 *
 * Every part is tested and the nearest wins, which is what makes a head shot a
 * head shot rather than whichever limb happened to be listed first — and why
 * the parts must be the same ones the renderer draws.
 */
export function traceShot(origin: Vec3, aim: Vec3, targets: readonly ShotTarget[]): Trace {
  let closest = MAX_SHOT_DISTANCE;

  for (const occluder of OCCLUDERS) {
    const distance = rayVsBox(origin, aim, occluder, closest);
    if (distance !== null && distance < closest) {
      closest = distance;
    }
  }

  let hitSeat: Seat | null = null;
  let hitPart: BodyPart | null = null;
  for (const target of targets) {
    const distance = rayVsOrientedBox(origin, aim, target.box, closest);
    if (distance !== null && distance < closest) {
      closest = distance;
      hitSeat = target.seat;
      hitPart = target.part;
    }
  }

  return {
    hitSeat,
    hitPart,
    endpoint: {
      x: origin.x + aim.x * closest,
      y: origin.y + aim.y * closest,
      z: origin.z + aim.z * closest,
    },
    distance: closest,
  };
}

File: server/nakama/arena/ray.go
package arena

// Where a shot goes, and what it finds.
//
// This is in the shared rules rather than on the server alone, and that was a
// deliberate change of mind. Hit registration is the most disputed mechanic in
// any shooter; leaving it as the one rule the conformance vectors did not pin
// would have been leaving out the part most worth pinning.

// RayVsBox reports how far along the ray it enters the box, and whether it ever
// does.
//
// The classic form of this multiplies by 1/direction and lets the infinities
// sort themselves out. That form produces 0 × ∞ = NaN for a ray running
// parallel to a face and starting on its plane, and Go and TypeScript do not
// agree about how a NaN travels through the comparisons that follow. So a
// degenerate axis is branched on and never divided through.
//
// A ray starting inside the box enters it at zero.
func RayVsBox(origin, direction Vec3, box Bounds, maxDistance float64) (float64, bool) {
	enter := 0.0
	exit := maxDistance

	// slab narrows the interval the ray is inside the box on one axis, and
	// reports whether anything is left of it.
	slab := func(originAxis, directionAxis, minAxis, maxAxis float64) bool {
		if directionAxis == 0 {
			return originAxis >= minAxis && originAxis <= maxAxis
		}
		a := (minAxis - originAxis) / directionAxis
		b := (maxAxis - originAxis) / directionAxis
		near, far := a, b
		if b < a {
			near, far = b, a
		}
		if near > enter {
			enter = near
		}
		if far < exit {
			exit = far
		}
		return enter <= exit
	}

	if !slab(origin.X, direction.X, box.MinX, box.MaxX) {
		return 0, false
	}
	if !slab(origin.Y, direction.Y, box.MinY, box.MaxY) {
		return 0, false
	}
	if !slab(origin.Z, direction.Z, box.MinZ, box.MaxZ) {
		return 0, false
	}

	return enter, true
}

// RayVsOrientedBox reports how far along the ray it enters a box with an
// orientation of its own.
//
// The same slab test, done in the box's frame instead of the world's. A part of
// a body carries its own right, up and forward, so the ray's origin and
// direction are projected onto those three axes with dot products and the
// result is an axis-aligned problem again. No matrix, no inverse, and nothing
// that is not a multiply and an add.
func RayVsOrientedBox(origin, direction Vec3, box OrientedBox, maxDistance float64) (float64, bool) {
	dx := origin.X - box.Centre.X
	dy := origin.Y - box.Centre.Y
	dz := origin.Z - box.Centre.Z
	along := func(axis Vec3) float64 {
		return float64(dx*axis.X) + float64(dy*axis.Y) + float64(dz*axis.Z)
	}
	cast := func(axis Vec3) float64 {
		return float64(direction.X*axis.X) + float64(direction.Y*axis.Y) + float64(direction.Z*axis.Z)
	}

	return RayVsBox(
		Vec3{X: along(box.Right), Y: along(box.Up), Z: along(box.Forward)},
		Vec3{X: cast(box.Right), Y: cast(box.Up), Z: cast(box.Forward)},
		Bounds{
			MinX: -box.Half.X, MinY: -box.Half.Y, MinZ: -box.Half.Z,
			MaxX: box.Half.X, MaxY: box.Half.Y, MaxZ: box.Half.Z,
		},
		maxDistance,
	)
}

// OrientedBox is a box that is not aligned to the world: a limb, a torso, a
// rifle.
type OrientedBox struct {
	Centre  Vec3
	Half    Vec3
	Right   Vec3
	Up      Vec3
	Forward Vec3
}

// ShotTarget is one part of one body, where it stood at the judged moment.
type ShotTarget struct {
	Seat string
	Part string
	Box  OrientedBox
}

// Trace is what one shot found.
type Trace struct {
	// HitSeat is who was hit, or empty for nobody.
	HitSeat string
	// HitPart is which part of them, which is what decides the damage.
	HitPart string
	// Endpoint is where the shot stopped: a body, a wall, or the end of its
	// range.
	Endpoint Vec3
	Distance float64
}

// TraceShot fires one shot and reports what it found.
//
// Walls are tested before bodies and win ties, so a target standing exactly
// against the far side of a crate is behind it rather than in front of it.
// Targets are given as boxes rather than as bodies because the server hands
// over where they were when the shooter saw them, not where they are now.
//
// Every part is tested and the nearest wins, which is what makes a head shot a
// head shot rather than whichever limb happened to be listed first — and why
// the parts must be the same ones the renderer draws.
func TraceShot(origin, aim Vec3, targets []ShotTarget) Trace {
	closest := MaxShotDistance

	for _, occluder := range Occluders {
		distance, hit := RayVsBox(origin, aim, boundsOf(occluder), closest)
		if hit && distance < closest {
			closest = distance
		}
	}

	hitSeat, hitPart := "", ""
	for _, target := range targets {
		distance, hit := RayVsOrientedBox(origin, aim, target.Box, closest)
		if hit && distance < closest {
			closest = distance
			hitSeat = target.Seat
			hitPart = target.Part
		}
	}

	return Trace{
		HitSeat: hitSeat,
		HitPart: hitPart,
		Endpoint: Vec3{
			X: origin.X + float64(aim.X*closest),
			Y: origin.Y + float64(aim.Y*closest),
			Z: origin.Z + float64(aim.Z*closest),
		},
		Distance: closest,
	}
}

func boundsOf(box Box) Bounds {
	return Bounds{
		MinX: box.MinX, MinY: box.MinY, MinZ: box.MinZ,
		MaxX: box.MaxX, MaxY: box.MaxY, MaxZ: box.MaxZ,
	}
}
